gapis/resolve/dependencygraph2/graph_visualization/graph_algorithms.go

```diff
@@ -152,3 +158,110 @@ func (g *graph) joinNodesWithZeroDegree() {
+		}
+	}
+}
+
+type chunk struct {
+	idToPosition       map[int]int
+	positionToLevelsID [][]int
+	done               bool
+}
+
+func assignLevelsIDToChunk(leftPosition, rightPosition int, currentLevelsID *[]int, positionToLevelsID *[][]int) {
+	if (rightPosition - leftPosition + 1) <= MAXIMUM_NUMBER_OF_NODES_BY_LEVEL {
+		for i := leftPosition; i <= rightPosition; i++ {
+			(*positionToLevelsID)[i] = make([]int, len(*currentLevelsID))
+			copy((*positionToLevelsID)[i], *currentLevelsID)
+		}
+	} else {
+		*currentLevelsID = append(*currentLevelsID, 1)
+		chunkSize := (rightPosition - leftPosition + 1 + MAXIMUM_NUMBER_OF_NODES_BY_LEVEL - 1) / MAXIMUM_NUMBER_OF_NODES_BY_LEVEL
+		newLeftPosition := leftPosition
+		newRightPosition := newLeftPosition + chunkSize - 1
+		chunkId := 1
+		for newLeftPosition <= rightPosition {
+			if newRightPosition > rightPosition {
+				newRightPosition = rightPosition
+			}
+			(*currentLevelsID)[len(*currentLevelsID)-1] = chunkId
+			assignLevelsIDToChunk(newLeftPosition, newRightPosition, currentLevelsID, positionToLevelsID)
+			newLeftPosition = newRightPosition + 1
+			newRightPosition = newLeftPosition + chunkSize - 1
+			chunkId++
+		}
+		*currentLevelsID = (*currentLevelsID)[:len(*currentLevelsID)-1]
+	}
+}
+
+func makeChunks(nodes []*node) {
+
+	labelAsAStringToChunk := map[string]*chunk{}
+	for _, currentNode := range nodes {
+		var labelAsAString bytes.Buffer
+		for i, name := range currentNode.label.LevelsName {
+			id := currentNode.label.LevelsID[i]
+			labelAsAString.WriteString(name)
+			currentChunk, ok := labelAsAStringToChunk[labelAsAString.String()]
+			if !ok {
+				currentChunk = &chunk{idToPosition: map[int]int{}}
+				labelAsAStringToChunk[labelAsAString.String()] = currentChunk
+			}
+			if _, ok := currentChunk.idToPosition[id]; !ok {
+				currentChunk.idToPosition[id] = len(currentChunk.idToPosition)
+			}
+			fmt.Fprintf(&labelAsAString, "%d/", id)
+		}
+	}
+
+	for _, currentNode := range nodes {
+		var labelAsAString bytes.Buffer
+		newLabel := &api.Label{}
+		for i, name := range currentNode.label.LevelsName {
+			id := currentNode.label.LevelsID[i]
+			labelAsAString.WriteString(name)
+			currentChunk := labelAsAStringToChunk[labelAsAString.String()]
+			chunkSize := len(currentChunk.idToPosition)
+			if chunkSize >= MINIMUM_NUMBER_OF_NODES_TO_BE_CHUNK {
+				if !currentChunk.done {
+					currentChunk.done = true
+					currentChunk.positionToLevelsID = make([][]int, chunkSize)
+					currentLevelsID := make([]int, 1)
+					assignLevelsIDToChunk(0, chunkSize-1, &currentLevelsID, &currentChunk.positionToLevelsID)
+				}
+				position := currentChunk.idToPosition[id]
+				newLevelsID := currentChunk.positionToLevelsID[position]
+				for _, id := range newLevelsID {
+					newLabel.PushBack(SUPER+name, id)
+				}
+			}
+			newLabel.PushBack(name, id)
+			fmt.Fprintf(&labelAsAString, "%d/", id)
+		}
+		currentNode.label = newLabel
+	}
+}
+
+func (g *graph) joinNodesInChunksByFrame() {
```

Comment:
Is there a reason that joining nodes by frame, and making the chunks have to be done together like this? Could we have separate `joinNodesByFrame()` and `makeChunks()` steps?

Reply:
I separated them in two different functions, join joinNodesByFrame() and makeChunks(), in addition I added the config for the global variables used in the chunk algorithm.